### PR TITLE
Add a scale op, related instances and examples

### PR DESCRIPTION
**Pull request by geyyer**
- Add element op
- Add instances
- Add example
- Add client example

### CODE REVIEW

File: client_example/32_convnd_fwd_convscale/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+if((DTYPES MATCHES "fp8") OR NOT DEFINED DTYPES)
```
Comment (bartekxk) [inline client_example/32_convnd_fwd_convscale/CMakeLists.txt]: Can you move it to client_example/24_grouped_conv_activation/grouped_convnd_fwd_scale ?

File: client_example/32_convnd_fwd_convscale/common.hpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline client_example/32_convnd_fwd_convscale/common.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
+// SPDX-License-Identifier: MIT
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]: Can you move it to example/62_convnd_activ ?

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: example/65_grouped_conv_fwd_convscale/run_convnd_fwd_convscale_example.inc [hunk 5]
```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/run_convnd_fwd_convscale_example.inc]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    {
+        throw std::runtime_error(
+            "wrong! device_conv with the specified compilation parameters does "
+            "not support this Conv problem");
+    }
+
+    float avg_time = invoker.Run(argument, StreamConfig{nullptr, time_kernel});
+
+    std::size_t ds_size   = 3; // 3 element-wise scale multipliers
+    std::size_t flop      = GetFlops<NDimSpatial>(e_g_n_k_wos_lengths, b_g_k_c_xs_lengths, ds_size);
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]: We should take into account elementwise flops

File: client_example/32_convnd_fwd_convscale/common.hpp [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    std::size_t in_mem_size  = GetInputByte<InDataType, NumDimSpatial>(in_lengths);
+    std::size_t wei_mem_size = GetWeightByte<WeiDataType, NumDimSpatial>(wei_lengths);
+    std::size_t d0_mem_size  = GetOutputByte<D0DataType, NumDimSpatial>(d0_lengths);
+    std::size_t d1_mem_size  = GetOutputByte<D1DataType, NumDimSpatial>(d1_lengths);
+    std::size_t d2_mem_size  = GetOutputByte<D2DataType, NumDimSpatial>(d2_lengths);
+    std::size_t out_mem_size = GetOutputByte<OutDataType, NumDimSpatial>(out_lengths);
+
+    SimpleDeviceMem in(in_mem_size);
+    SimpleDeviceMem wei(wei_mem_size);
+    SimpleDeviceMem d0(d0_mem_size);
```
Comment (bartekxk) [inline client_example/32_convnd_fwd_convscale/common.hpp]: It should be enough to allocate memory only for 1 element and pass {N, Do, Ho, Wo, G, K} lengths and {0, 0,0, 0, 0, 0} strides for broadcast.

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp [hunk 8]
```diff
@@ -528,6 +528,26 @@ struct UnaryTypeConvert<ck::bhalf_t, float>
     }
 };
 
+struct ConvScale
```
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/element/element_wise_operation.hpp]: You can add next operator in  include/ck/tensor_operation/gpu/element/combined_element_wise_operation.hpp?
for example `QuadrupleWithUnaryCombinedOp` and use `QuadrupleWithUnaryCombinedOp<Multiply, Multiply,Multiply, PassThrough, PassThrough,PassThrough>`

But it looks a little bit complicated so feel free to skip this comment and add new Elementwise Op :)

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp (hunk 6 above)
Comment (geyyer) [reply example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]: Sure, we take into account, here is how we calculate it in GetFlops: G * N * C * output spatial lengths product * (2 * K * filter spatial lengths product + number of scale factors). Overall, elementwise op adds ds_size*output_pixels flops.

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp (hunk 8 above)
Comment (geyyer) [reply include/ck/tensor_operation/gpu/element/element_wise_operation.hpp]: Wow, meta-elementwise-op, I haven't thought about it! Let's keep it simple so far and once we add all required elementwise ops, we can generalize with a meta-op, what do you think?

Comment (bartekxk) [reply include/ck/tensor_operation/gpu/element/element_wise_operation.hpp]: Sure!

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                                                  conv_param.input_left_pads_,
+                                                  conv_param.input_right_pads_,
+                                                  in_element_op,
+                                                  wei_element_op,
+                                                  PassThrough{});
+
+        ref_invoker.Run(ref_argument);
+
+        out_host.ForEach([&](auto&, auto idx) {
+            out_element_op(out_host(idx), c(idx), d0(idx), d1(idx), d2(idx));
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]:
```suggestion
            out_element_op(out_host(idx), c(idx), d0(0), d1(0), d2(0));
```

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_xdl_convscale_fp8.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        S<1, 0, 2>,  // BBlockTransferThreadClusterArrangeOrder
+        S<1, 0, 2>,  // BBlockTransferSrcAccessOrder
+        2,           // BBlockTransferSrcVectorDim
+        8,           // BBlockTransferSrcScalarPerVector
+        8,           // BBlockTransferDstScalarPerVector_BK1
+        1,           // BBlockLdsExtraN
+        1,
+        1,
+        S<1, 32, 1, 8>,
+        8,
```
Comment (bartekxk) [inline example/65_grouped_conv_fwd_convscale/convnd_fwd_xdl_convscale_fp8.cpp]:
```suggestion
        1,
```
If D0, D1 and D2 are scalars then C scalar per vector must be 1. It is bad limitation so maybe we should consider some other solution.

File: library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+#ifdef CK_ENABLE_FP8
+        // generic instance
+        DeviceGroupedConvFwdMultipleABD_Xdl_CShuffle<NDimSpatial,ALayout,BLayout,    DsLayout,ELayout,   F8,   F8,     F32,      F32,  Tuple<F32, F32, F32>,    F8, PassThrough, PassThrough,     ConvScale,       ConvSpec, GemmMNKPadding,        1,    64,    64,    64,    32,   8,   8,   32,   32,    2,    2,     S<4, 16, 1>,     S<1, 0, 2>,     S<1, 0, 2>,              2,              1,              8,         1,     S<4, 16, 1>,     S<1, 0, 2>,     S<1, 0, 2>,             2,              1,              8,         1,           1,           1,               S<1, 16, 1, 4>,               1,     F8,      F8>,
```
Comment (bartekxk) [inline library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp]: With actual implementation all of these instances should have CScalarPerVector equal to 1 due to D0, D1 D2 are scalars.

**Comment by bartekxk** [comment]
I have good idea for this problem:

- Add struct SCALAR : public BaseTensorLayout
- Pass Tuple<SCALAR, SCALAR, SCALAR> as layouts for Ds
- Pass only one value with shape ={1}, strides={0} for each d0,d1,d2
- ⁠Add private scale0, scale1, scale2 members in ConvScale elementwise, add some __device__ function SetScales(scale0, scale1, scale2) and run it in __global__ function
- in kernel add some if constexpr(is_same_v<Ds_layout.AT(i), SCALAR>), in this case set scale0, scale1, scale2 members to values under d0, d1, d2 pointer and pass empty D tuple to gridwise gemm

Pros of this solution:

- Easier API for the user (he dont need to pass some weird strides)
- Less memory usage (only need to allocate sizeof(datatype) bytes for each d0, d1,d2
- Kernel dont need to load d0, d1, d2 for each C store
- It allows to use CScalarPerVector > 1

File: example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp (hunk 3 above)
Comment (geyyer) [reply example/65_grouped_conv_fwd_convscale/convnd_fwd_convscale_common.hpp]: Fixed in [4e41656](https://github.com/ROCm/composable_kernel/pull/1242/commits/4e41656aa8731b530422b08b57a5edabe8453eb5)

File: client_example/32_convnd_fwd_convscale/CMakeLists.txt (hunk 1 above)
Comment (geyyer) [reply client_example/32_convnd_fwd_convscale/CMakeLists.txt]: Fixed in [d68172d](https://github.com/ROCm/composable_kernel/pull/1242/commits/d68172d9ac76c40f4d5d5e30539e1f7f5cfc81f5)

File: example/62_convnd_activ/convscale/run_convnd_fwd_convscale_example.inc [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        conv_param = ck::utils::conv::parse_conv_param(num_dim_spatial, 5, argv);
+    }
+
+    // sample scale values, same as host
+    float scale_in  = 0.1f;
+    float scale_wei = 0.2f;
+    float scale_out = 0.3f;
+
+    // allocate device memory for scales
+    DeviceMem scale_in_device_buf(sizeof(float));
```
Comment (geyyer) [inline example/62_convnd_activ/convscale/run_convnd_fwd_convscale_example.inc]: Here is the way we handle scale pointers on device

File: example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+
+        ref_invoker.Run(ref_argument);
+
+        // sample scale values, same as device
+        float scale_in_host  = 0.1f;
+        float scale_wei_host = 0.2f;
+        float scale_out_host = 0.3f;
+
+        // set up pointers to pass to out_element_op_host
+        float* scale_in_host_ptr  = &scale_in_host;
```
Comment (geyyer) [inline example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp]: And here is the way we can handle it on the host side

File: example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+    copy(wei_g_k_c_xs_desc.GetStrides(), b_g_k_c_xs_strides);
+    copy(out_g_n_k_wos_desc.GetLengths(), e_g_n_k_wos_lengths);
+    copy(out_g_n_k_wos_desc.GetStrides(), e_g_n_k_wos_strides);
+    copy(conv_param.conv_filter_strides_, conv_filter_strides);
+    copy(conv_param.conv_filter_dilations_, conv_filter_dilations);
+    copy(conv_param.input_left_pads_, input_left_pads);
+    copy(conv_param.input_right_pads_, input_right_pads);
+
+    // sample scale values, same as host
+    float scale_in  = 0.1f;
```
Comment (zjing14) [inline example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp]: Should we change scales per every run based on your example design?

File: library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+#ifdef CK_ENABLE_FP8
+        // generic instance
+        DeviceGroupedConvFwdMultipleABD_Xdl_CShuffle<NDimSpatial,ALayout,BLayout,  DsLayout,ELayout,    F8,    F8,     F32,      F32,  Tuple<>,    F8, PassThrough, PassThrough,     ConvScale,       ConvSpec, GemmMNKPadding,        1,    64,    64,    64,    32,   8,   8,   32,   32,    2,    2,     S<4, 16, 1>,     S<1, 0, 2>,     S<1, 0, 2>,              2,              1,              8,         1,     S<4, 16, 1>,     S<1, 0, 2>,     S<1, 0, 2>,             2,              1,              8,         1,           1,           1,               S<1, 16, 1, 4>,               1,     F8,      F8>,
```
Comment (zjing14) [inline library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp]: Suggestion: We can make OutElementOps (i.e., ConvScale) as a template argument to reuse instances for other fused cases.

File: example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp (hunk 14 above)
Comment (geyyer) [reply example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp]: Yes, that's the main point, on each iteration we scale fp8 output into fp32. And the scale is derived on the previous layer, so they may change every iteration.

File: library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp (hunk 15 above)
Comment (geyyer) [reply library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp]: Good point, I'll refactor it as a template argument!

Comment (geyyer) [reply library/include/ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_convscale_instance.hpp]: Added in [c63d3f1](https://github.com/ROCm/composable_kernel/pull/1242/commits/c63d3f10595f59908b8471d1c103e5c116f20151)

File: example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp (hunk 14 above)
Comment (geyyer) [reply example/62_convnd_activ/convscale/convnd_fwd_convscale_common.hpp]: Added random values in [0dfd142](https://github.com/ROCm/composable_kernel/pull/1242/commits/0dfd142a89ef62be55ceb8d3daaa5f8b1829ee17)

**Review by zjing14 — APPROVED** [review]

**Pull request merged**